### ISSUE TITLE
Investigate making compressed references a runtime and not compile time decision

### ISSUE THREAD

**Issue by charliegracie** [eclipse/openj9#643]
Currently the OpenJ9 VM can not support both compressed references and not compressed references for 64 bit JVMs at the same time.  This means to create a JDK that supports both you need to have 2 JVMs which significantly increases the footprint of the JDK package, the complexity of creating the JDK and the time it takes to create the JDK.

I am creating this Issue to start the conversation about what needs to be completed to make compressed references a runtime decision such that the JVM supports both at the same time.

**Comment by charliegracie** [comment]
My initial thoughts are as follows:
1. We likely need to compile the interpreter multiple times and pick the right one based on whether or not compressed references is enabled.
2. The JIT already queries the GC for most of it's compressed references decisions so I am hoping it is not that hard to complete the extra work there.
3. The GC will likely have to deal with minor performance regressions for tight loops reading / writing object references
4. Currently there are 3 (maybe more) compile flags that control whether or not compressed references can be used.  I would suggest we move all of this code under a single compiler flag to make the rest of these changes easier.

**Comment by andrewcraik** [comment]
Whatever mechanism we design for the JIT to query the compressed reference configuration - we need to make sure that it is very cheap to run - the JIT should not have to call over the VM every time to get the state otherwise we will see a compile-time penalty that could become prohibitive. Whether we do this by caching the VM result in the JIT to make it easier to inline or something I don't know, but just a point for consideration.

I also assume this item is to allow it to be a runtime decision, but a decision taken once at the start of a given VM so that we don't have to worry about it changing "on the fly"?

**Comment by charliegracie** [comment]
This would be something decided on startup that can not change at runtime after we have initialized the JVM.  I agree this would be something that could be cached in the JIT or any other component for speed purposes.

**Comment by ymanton** [comment]
> 2. The JIT already queries the GC for most of it's compressed references decisions so I am hoping it is not that hard to complete the extra work there.

The assembly routines that use the compile-time flags will be the most problematic to deal with.

**Comment by JamesKingdon** [comment]
How much work would it take to get a prototype running so that we can assess the performance impact?

**Comment by mstoodle** [comment]
Only 51 occurrences of *COMPRESSED* in the JIT that are in assembler files. Anything else I believe could just be taught to use a cached query result so I'm not particularly worried about them.

Of those 51, 10 are && J9VM_OPT_TENANT so presumably that's just dead code...

z seems to have the most occurrences(21), then p (11), then x (8) . But there are uses in macros that then spread throughout the rest of the file, so these numbers may be somewhat misleading.

**Comment by mstoodle** [comment]
From the JIT side, I would like to start moving forward with this work...I don't think it's a lot of work, but we can make progress on it even in the absence of the VM work to make the driving query truly dynamic. We'll try to wrangle up some people to start moving forward here.

**Comment by charliegracie** [comment]
The first step on the VM/GC side would be to consolidate all of the build flags into a single build flag. Once this work is complete then it will be easier to design out what changes need to be made to make the decision at runtime. This work will need to be completed in the OMR repo as well since it knows about compressed references. 

@mstoodle as part of the JIT work could the same build flag cleanup happen as well?

Initially ALL of the flags should stay enabled while code is being moved under the single build flag. Once all of the code is moved then we can disable / remove the other build flags.

**Comment by mstoodle** [comment]
Yeah, I forgot to mention the multiple build flags because only `J9VM_INTERP_COMPRESSED_OBJECT_HEADER` is being used in the assembler files I was focusing on.

There's a lot of cpp/hpp code dependent on `J9VM_GC_COMPRESSED_POINTERS`. I don't see any reason this couldn't be unified on a single thing assuming we would never independently use compressed references without compressing the object header....?

For the JIT, the difference between one and two dynamic queries is negligible.

**Comment by charliegracie** [comment]
Before we begin though I think there are some design decisions we should make.

In the future should OpenJ9 still be able to build a compressed references only build? ie not a runtime decision?  Should OpenJ9 be able to build a full 64 bit (non compressed) only build? 

Depending on the answers to the above questions the solution may be more or less complicated to complete.

**Comment by mstoodle** [comment]
Inside the JIT, I think we would move forward with two queries that would (at this point) simply be driven by #ifdef's on the two build flags (later to be transitioned to the actual cached dynamic test).

I'm not that worried about the additional library size if the build compiler cannot propagate the setting throughout the JIT code, though we've had to do that in other scenarios in the past (e.g. the (n)ever popular and now vanquished query isSmall() ). But outside the JIT (and possibly in some of our assembler sequences, depending on how we do this work), I could imagine the question being much more relevant to avoid the performance impact of the dynamic test.

I reserve the right to reverse that "worried" statement once we do the work :) .

**Comment by mstoodle** [comment]
How much impact do we think these dynamic tests will have on large heap scenarios where compressed references aren't viable? OpenJDK doesn't seem to believe that's an important case to optimize. Do we think the delta will be more than a few percent on applications that fall into this category?

**Comment by DanHeidinga** [comment]
The impact will be largest during startup and during a GC cycle.

The startup cost can likely be alleviated by compiling 2 additional copies of the BytecodeInterpreter:
* compressedrefs
* compressedrefs debug
* default
* default debug
and using a value in the J9JavaVM to determine which interpreter to run. 

To get a better handle on the startup cost we'll need to look at everywhere we differentiate between a j9object and an fj9object.

@charliegracie Any thoughts on where costs would hit GC?

**Comment by charliegracie** [comment]
It is likely going to be measurable in GC times with the simple solution of adding a runtime check. More complicated solutions likely involve compiling some of the GC infrastructure twice and depending on the VM option pick the right one. Depending on my question above in this [comment](https://github.com/eclipse/openj9/issues/643#issuecomment-446231718) the difficulty of these changes is quite different.

**Comment by vijaysun-omr** [comment]
@mpirvu FYI

**Comment by JamesKingdon** [comment]
What would we need to do to be able to write small amounts of code for the VM impl that could be compiled by the JIT? I guess we wouldn't want to write the code in Java or bytecodes, so a C to bytecode compiler would be needed? It's probably not practical, but it feels like a neat solution to these sorts of dynamic problems that need late optimization.

**Comment by zl-wang** [comment]
where are we with the status of this issue?
there are some urgency build-up regarding a single driver of OpenJ9 (no matter which way) to be used in PoC(s) such as ServiceNow and Medallia.

**Comment by DanHeidinga** [comment]
See also #5412 where @gacholio is starting to investigate making compressed / not a runtime decision

**Comment by fjeremic** [comment]
@gacholio do we have JIT resources lined up to help with this effort? Do you need help on the JIT side?

**Comment by gacholio** [comment]
@fjeremic No, I expect to do the JIT work myself.
